### PR TITLE
Separate `depcode.py` into three files

### PR DESCRIPTION
**Pull request by yardasol**
## Summary of changes

The purpose of this PR is to make the changes in #172 easier to see for reviewers.

Specifically, this PR does the following:
- Renames the depcode classes:
  - `DepcodeSerpent` -> `SerpentDepcode`
  - `DepcodeOpenMC` -> `OpenMCDepcode`
- Separates the `depcode.py` file into three difference files:
  - `abc.py`, storing `Depcode`
  - `openmc_depcode.py`, storing `OpenMCDepcode`
  - `serpent_depcode.py`, storing `SerpentDepcode`

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Required for Merging
- [x] I have read the **CONTRIBUTING** document.
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
  - [x] I have updated the documentation accordingly. 
- [x] My change is a source code change
  - [x] I have added/modified tests to cover my changes
  - [ ] I have explained why my change does not require new/modified tests
- [x] My change is a user-facing change
  - [x] I have recorded my changes in the changelog for the upcoming release
- [ ] My change is exclusively related to the repository (e.g. GitHub actions workflows, PR/issue templates, etc.)
  - [ ] I have verified or justified that my change will work as intended.
- [x] All new and existing tests passed.
  - [x] CI tests pass
  - [x] Local tests pass (including Serpent2 integration tests)
## Associated Issues and PRs

- Issue: #172 

## Associated Developers

- Dev: @munkm 

## Checklist for Reviewers

Reviewers should use [this link](https://arfc.github.io/manual/guides/pull_requests) to get to the 
Review Checklist before they begin their review.

### CODE REVIEW

File: doc/releasenotes/v0.5.0.rst
Comment (abachma2) [inline doc/releasenotes/v0.5.0.rst:124]: Is there any reason you have one with the arrow and one without? Is `OpenMCDepcode` new? I thought you were renaming it.

File: saltproc/abc.py
Comment (abachma2) [inline saltproc/abc.py:3]: Couple things:
- are you importing the ABC class from this module into this module?
- Did you name this class `ABC` as in "abstract class"? Is this class ever seen by the user? If it is then this feels like a not-so intuitive class name for a user to have to remember. If the user doesn't see this, then disregard.

File: saltproc/openmc_depcode.py
Comment (abachma2) [inline saltproc/openmc_depcode.py:80]: Do you mean to leave these paths hardcoded? Wouldn't you want to read in the values of the template_input_file dictionary?

File: saltproc/openmc_depcode.py
Comment (abachma2) [inline saltproc/openmc_depcode.py:85]: I'm assuming that most of the functions are placeholders for now, correct?

File: saltproc/abc.py
Comment (samgdotson) [inline saltproc/abc.py:1]: This file should be renamed to avoid a potential namespace conflict with `abc`. Also so that it's clearer to users what lives here.

Be sure to use `git mv abc.py some-other-name.py` to minimize the number of files in the commit.

File: tests/integration_tests/run_no_reprocessing/test.py
Comment (abachma2) [inline tests/integration_tests/run_no_reprocessing/test.py:49]: Why are you saying `almost equal` here? Is there some rounding error between values of the array here?

**Review by abachma2 — COMMENTED** [review]
Hi @yardasol. Great work on this, I'm sure if took you some time to make sure you found all of the variable names that needed to updated. I have a few questions about some of the files. I will approve once you address my comments, although they are pretty minor.

File: saltproc/abc.py
Comment (samgdotson) [reply saltproc/abc.py:3]: This pattern indicates that `Depcode` is an abstract class -- i.e. it inherits from the `abc.ABC` class. The purpose of abstract classes is to provide a template for future class creation. The expected usage is something like

```py 
MyClass(Depcode)

def read_depcode_info():
  do stuff
```

where `MyClass` overrides the `read_depcode_info` method (which doesn't currently do anything).

File: saltproc/app.py
Comment (samgdotson) [inline saltproc/app.py:243]: I guess this is fine, lol. Why rearrange?

File: saltproc/openmc_depcode.py
Comment (samgdotson) [inline saltproc/openmc_depcode.py:14]: @abachma2 example of `Depcode` 😄

File: saltproc/openmc_depcode.py
Comment (samgdotson) [reply saltproc/openmc_depcode.py:85]: This actually confuses me, too. These are no longer `@abstractmethods` so shouldn't they do something?

File: tests/conftest.py
Comment (samgdotson) [inline tests/conftest.py:40]: I don't love this... I get that the first several variables are unused -- but I think they should still have proper names. I don't think there are any performance benefits to this either. This is just a preference though, so if you have a compelling reason to use underscores go ahead, I suppose.

File: saltproc/abc.py
Comment (samgdotson) [inline saltproc/abc.py:69]: Are these files created dynamically?

File: saltproc/openmc_depcode.py
Comment (samgdotson) [reply saltproc/openmc_depcode.py:80]: I agree with this, it seems restrictive to have these hardcoded.

File: saltproc/openmc_depcode.py
Comment (samgdotson) [inline saltproc/openmc_depcode.py:152]: Just an FYI using subprocess.STDOUT can create some challenges for users on different platforms. 

I had problems with it this summer when using `watts`. https://github.com/watts-dev/watts/pull/52.

Comment (samgdotson) [reply saltproc/openmc_depcode.py:152]: Namely users with
- WSL2
- windows

**Review by samgdotson — CHANGES_REQUESTED** [review]
Good job @yardasol. There are some minor-ish changes. In particular I think your file names could be clearer. `abc` and `app` are not useful indicators of their contents.

File: saltproc/abc.py
Comment (abachma2) [reply saltproc/abc.py:3]: I get the `Depcode` being an abstract class part. My confusion is where the `ABC` class is coming from. Is `ABC` is being imported from the `abc` module into an `abc` module? If so, then is it another `abc` module this is being pulled from? Or is the `abc` module it's importing from a python package, like numpy or pandas?

Comment (samgdotson) [reply saltproc/abc.py:3]: He's importing from the latter. My suggestion was to rename the file to avoid this confusion.

Comment (abachma2) [reply saltproc/abc.py:3]: I support this suggestion. Clearly I was confused by this.

File: doc/releasenotes/v0.5.0.rst
Comment (yardasol) [reply doc/releasenotes/v0.5.0.rst:124]: `OpenMCDepcode` doesn't exist in v0.4.0 yet, so there's no need to indicate that the name has been changed inside the development period between v0.4.0 and v0.5.0.

File: saltproc/abc.py
Comment (yardasol) [reply saltproc/abc.py:3]: I picked `abc.py` as it can hold many different classes that use abstract functions, but perhaps this isn't the best way of doing things.

File: saltproc/openmc_depcode.py
Comment (yardasol) [reply saltproc/openmc_depcode.py:85]: These are placeholders and will be implemented in a future PR in the near future. Right now these methods don't actually do anything.

Comment (yardasol) [reply saltproc/openmc_depcode.py:85]: These are placeholders and will be implemented in a future PR in the near future. Right now these methods don't actually do anything.

File: tests/integration_tests/run_no_reprocessing/test.py
Comment (yardasol) [reply tests/integration_tests/run_no_reprocessing/test.py:49]: I use this because monte-carlo codes are stochastic and can have slightly different results each time you run them, which would cause a `assert_equal` statement to fail.

File: saltproc/app.py
Comment (yardasol) [reply saltproc/app.py:243]: Personal preference, as well as consistency with how we would refer to these in natural language (i,e, Serpent depletion code becomes SerpentDepcode rather than DepcodeSerpent)

File: tests/conftest.py
Comment (yardasol) [reply tests/conftest.py:40]: My perspective is that they are not relevant to anything else in the function, and giving them a null-like name indicates that. Maybe @munkm has some perspective on this?

File: doc/releasenotes/v0.5.0.rst
Comment (abachma2) [reply doc/releasenotes/v0.5.0.rst:124]: Ok. That makes sense.

File: saltproc/abc.py
Comment (abachma2) [reply saltproc/abc.py:3]: If that's the case, maybe just write out `abstract_classes.py`?

File: tests/integration_tests/run_no_reprocessing/test.py
Comment (abachma2) [reply tests/integration_tests/run_no_reprocessing/test.py:49]: So this test is based on some result out of Serpent or OpenMC then? If that's the case, can you specify a random seed? Or does doing that still have some small variation in the result?

**Comment by abachma2** [comment]
> @samgdotson I agree we should rename `app.py`, but not in this PR. Is it okay if I rename `abc.py` in #172 just so that we can avoid any nasty merge conflicts?

I am okay with putting this into a different PR. I see that renaming the file is on the To Do list for that PR, and I like the name you suggest there in for `abc.py`. I do worry that #172 will be very large though, so please be aware of its size. Feel free to break it up into a few PRs.

File: saltproc/openmc_depcode.py
Comment (yardasol) [reply saltproc/openmc_depcode.py:152]: Thanks for the heads up. I'll look into fixing this as part of a PR that addresses #173

File: saltproc/abc.py
Comment (yardasol) [reply saltproc/abc.py:69]: Yes, and their names will be `inter_input` and `iter_mat` respectively. They are hardcoded because they don't really need to be seen by the user.

File: saltproc/openmc_depcode.py
Comment (yardasol) [reply saltproc/openmc_depcode.py:80]: They are hardcoded OpenMC by default expects that the settings, geometry, and material files are named `settings.xml`, `geometry.xml` `materials.xml` respectively. They are also hardcoded because they are modified instead of the user's original OpenMC input files. The user has no need to specify a name for them in the first place.

Comment (abachma2) [reply saltproc/openmc_depcode.py:80]: So the user can specify files that they created from any location, then the code will copy them to have these names and be in this location. Then the code continuously calls these copied files and edits them as needed? 

I get hardcoding the names, if that is what OpenMC requires. But would the file path ever change?

File: tests/integration_tests/run_no_reprocessing/test.py
Comment (yardasol) [reply tests/integration_tests/run_no_reprocessing/test.py:49]: The default sensitivity of `assert_almost_equal` is on the order of 10^-6, which I think is sufficiently small since we are comparing errors, not base values.

Comment (abachma2) [reply tests/integration_tests/run_no_reprocessing/test.py:49]: That makes sense. And this test passes?

**Comment by yardasol** [comment]
@samgdotson @abachma2 Is this good to be merged?

File: saltproc/openmc_depcode.py
Comment (yardasol) [reply saltproc/openmc_depcode.py:80]: > So the user can specify files that they created from any location, then the code will copy them to have these names and be in this location. Then the code continuously calls these copied files and edits them as needed?

Correct

> I get hardcoding the names, if that is what OpenMC requires. But would the file path ever change?

The file path will only change if there's a change to OpenMC that requires us to change it as such.

File: tests/integration_tests/run_no_reprocessing/test.py
Comment (yardasol) [reply tests/integration_tests/run_no_reprocessing/test.py:49]: Yes

**Review by abachma2 — APPROVED** [review]
I am approving this PR. I won't merge though until after @samgdotson gives his approval as well.

**Review by samgdotson — APPROVED** [review]
I guess this is fine if the issues raised will be resolved later.

**Pull request merged**
